Readme.md
# ICS Connectivity Agent Scripts



This Asset contains

- Blog Article available at ATeam Chronicles

  - http://www.ateam-oracle.com/ics-connectivity-agent-advanced-configuration/

- Sample Scripts
  - [Advanced Configuration Scripts](./advancedConfigScripts)
  - [Credential Configuration Scripts](./credentialScripts )

------



## ICS Connectivity Agent Advanced Configuration (Blog Snippet)

Oracle’s Integration Cloud Service (ICS) provides a feature that helps with the integration challenge of cloud to ground integrations with resources behind a firewall. This feature is called the ICS Connectivity Agent (additional details about the Agent can be found under [New Agent Simplifies Cloud to On-premises Integration](https://blogs.oracle.com/integration/entry/new_agent_simplifies_cloud_to)). The design of the Connectivity Agent is to provide a safe, simple, and quick setup for ICS to on-premise resources. In many cases this installation and configuration is an almost no-brainer activity. However, there are edge cases and network configurations that make this experience a bit more challenging.

We have encountered the following post-installation challenges with the ICS 16.3.5 Connectivity Agent:

1. Networks containing proxy server with SSL and/or Man In The Middle (MITM) proxy
2. On-premise resources requiring SSL
3. nonProxyHost required for on-premise resources
4. White list OMCS and upgrade URIs

It’s important to note that future releases of ICS may improve on these configuration challenges. However, some are not related to the product (e.g., network white list) and appropriate actions will need to be coordinated with the on-premise teams (e.g., network administrators).

### Import Certificates

One of the more challenging activities with post-configuration of the ICS Connectivity Agent is updating the keystore with Certificates that the agent needs to trust. Since the agent is a lightweight, single server, WebLogic installation, there are no web consoles available to help with the certificate import. However, if you investigate this topic on the internet you will eventually end up with details on using java keytool and WebLogic WLST to accomplish this task. Instead of doing all this research, I am including a set of scripts (bash and WLST) that can be used to expedite the process. The scripts are comprised of 4 files where each file contains a header that provides details on how the script works and its role in the process. Once downloaded, please review these headers to make yourself familiar with what is required and how they work together.

The following is a step-by-step example on using these scripts:

1. Download the scripts into a directory on the machine where the Connectivity Agent is running
2. Update the importToAgentEnv.sh to reflect your agent environment
3. Create a subdirectory that will be used to hold all the certificates that will need to be imported to the agent keystore
4. Download or copy all certificates in the chain to the directory created in the previous step
5. Execute the createKeystore.sh (NOTE: This script has created a file called importToAgent.ini that contains details that will be used by the importToAgent.py WLST script)
6. Make sure your agent server is running and execute the importToAgent.sh

At this point you will have imported the certificates into the keystore of the running Connectivity Agent. I always bounce the agent server to make sure it starts cleanly and everything is picked up fresh.

### For Complete Details, See Blog:

http://www.ateam-oracle.com/ics-connectivity-agent-advanced-configuration/


changelog.md
- Advanced Configuration Script History
  - Version 1.1 removed the Agent Password from the importToAgent.sh bash script and implemented a prompt
  
- Credential Configuration Scripts History
  - No Longer Needed ... for informational/historic purposes only


advancedConfigScripts/importToAgent.py
#===============================================================================
# Title:       importToAgent.py - ICS Connectivity Agent keystore importer
#
# Disclaimer:  This script is provided "AS IS" and without any official support
#              from Oracle. Its use needs to be performed using details from the
#              comments section and/or readme file (if one is included). Any bugs
#              encountered, feedback, and/or enhancement requests are welcome.
#
#              No liability for the contents of this script can be accepted. Use
#              the concepts, examples, and information at your own risk. However,
#              great care has been taken to ensure that all technical information
#              is accurate and as useful as possible.
#
#              Copyright © 2018, Oracle and/or its affiliates. All rights reserved.
#              The Universal Permissive License (UPL), Version 1.0
#
#
# Description: This script is used to import an existing keystore file (.jks)
#              into the keystore of a running instance of the ICS Connectivity
#              Agent.  It relies on the importToAgent.ini file (generated by
#              createKeystore.sh) that contains all the keystore details for a
#              successful import.  The following is an example of this .ini:
#
#              [ImportKeyStore]
#              appStripe: system
#              keystoreName: trust
#              keyAliases: root-VeriSignClass3-G5,intermediate-Symantec-G4,main-us
#              keyPasswords: changeit,changeit,changeit
#              keystorePassword: changeit
#              keystorePermission: true
#              keystoreType: JKS
#              keystoreFile: ./certificates/agentcerts.jks
#
#              If no command-line arguments are provided, the following defaults
#              will be used:
#                * Agent Server User:          weblogic
#                * Agent Server User Password: welcome1
#                * Agent Server Host Name:     localhost
#                * Agent Server Host Port:     7001
#
# Author:      Greg Mally
#
# Date:        31-October-2016
#
# Version:     1.0    
#
# Usage:       ./wlst.sh -u weblogic -p welcome1 --host localhost --port 7001
#
# Notes:       Although this script can be kicked off using wlst.sh directly,
#              it is recommended to use the importToAgent.sh.
#
#              After running this script it is recommended to bounce the
#              Connectivity Agent server to make sure everything loads as
#              expected.
#
#===============================================================================

import sys
import getopt
import ConfigParser
import java.lang.String as jstring
import java.lang.System as jsystem


# Usage function
def usage():
  print ''
  print 'NOTE: It is recommended to run this script using the importToAgent.sh'
  print 'Usage:       ./wlst.sh [-u AgentUserName] [-p AgentPassword] [-h,--host HostName] [-t,--port HostPort] [--help]'
  print 'Defaults:    ./wlst.sh -u weblogic -p welcome1 --host localhost --port 7001'
  print 'For Example, ./wlst.sh -u agentAdmin -p admin!'
  print ''


# Parse Config
config = ConfigParser.ConfigParser()
config.read('importToAgent.ini')


# Setup for processing command-line options
try:
  opts, args = getopt.getopt(sys.argv[1:], "u:p:h:t:",["host=", "port=", "help="])
except getopt.GetoptError, err:
  print str(err)
  usage()
  sys.exit(2)

  
# Default values
user="weblogic"
password="welcome1"
host="localhost"
port="7001"


# Assign parameters to variables
for opt, arg in opts:
  if opt == "-u":
    user = arg
  elif opt == "-p":
    password = arg
  elif opt in ("-h", "--host"):
    host = arg
  elif opt in ("-t", "--port"):
    port = arg
  elif opt == "--help":
    usage()
    sys.exit()
  else:
    print "Unknown options: ", opt
    print "Exiting WLST process ..."
    sys.exit()

# Setup connection string
connectStr = "t3://" + str(host) + ":" + str(port)

# Setup variables
APPSTRIPE           = config.get('ImportKeyStore', 'appStripe')
KEYSTORENAME        = config.get('ImportKeyStore', 'keystoreName')
KEYALIASES          = config.get('ImportKeyStore', 'keyAliases')
KEYPASSWORDS        = config.get('ImportKeyStore', 'keyPasswords')
KEYSTOREPASSWORD    = config.get('ImportKeyStore', 'keystorePassword')
KEYSTOREPERMISSION  = config.getboolean('ImportKeyStore', 'keystorePermission')
KEYSTORETYPE        = config.get('ImportKeyStore', 'keystoreType')
KEYSTOREFILE        = config.get('ImportKeyStore', 'keystoreFile')


print 'Using the following for the importKeyStore:'
print 'WebLogic URI        = ', connectStr
print 'WebLogic User       = ', user
print 'WebLogic Password   = ', password
print 'appStripe           = ', APPSTRIPE
print 'keystoreName        = ', KEYSTORENAME
print 'keyAliases          = ', KEYALIASES
print 'keyPasswords        = ', KEYPASSWORDS
print 'keystorePassword    = ', KEYSTOREPASSWORD
print 'keystorePermission  = ', config.get('ImportKeyStore', 'keystorePermission')
print 'keystoreType        = ', KEYSTORETYPE
print 'keystoreFile        = ', KEYSTOREFILE
print ''

try:
    connect(user,password,connectStr)

    serverRuntime()

    svc = getOpssService(name='KeyStoreService')

    svc.importKeyStore(appStripe=APPSTRIPE, name=KEYSTORENAME, aliases=KEYALIASES, keypasswords=KEYPASSWORDS, password=KEYSTOREPASSWORD, permission=KEYSTOREPERMISSION, type=KEYSTORETYPE, filepath=KEYSTOREFILE)

except Exception, ex:
    print 'Error importing keystore keys: '
    print str(ex)


advancedConfigScripts/createKeystore.sh
#!/bin/bash
#===============================================================================
# Title:       createKeystore.sh
#
# Disclaimer:  This script is provided "AS IS" and without any official support
#              from Oracle. Its use needs to be performed using details from the
#              comments section and/or readme file (if one is included). Any bugs
#              encountered, feedback, and/or enhancement requests are welcome.
#
#              No liability for the contents of this script can be accepted. Use
#              the concepts, examples, and information at your own risk. However,
#              great care has been taken to ensure that all technical information
#              is accurate and as useful as possible.
#
#              Copyright © 2018, Oracle and/or its affiliates. All rights reserved.
#              The Universal Permissive License (UPL), Version 1.0
#
#
# Description: This script is used to create a keystore ($JKS_FILENAME) in the
#              certificates directory (-cd=...) containing all the certificates
#              that match the file pattern provided on the command-line (-cp=).
#              The details needed later for the import process (importToAgent.py)
#              are captured in the importToAgent.ini by this script.  This is
#              the first script to run in a two step process for importing the
#              certificates into the ICS Connectivity Agent keystore.
#
# Author:      Greg Mally
#
# Date:        31-October-2016
#
# Version:     1.0    
#
# Usage:       ./createKeystore.sh -cd=./certificates -cp=*.crt
#
# Notes:       The above usage assumes that this script has execute permission.
#              If it doesn't have execute permission, then add "bash" in front
#              of the command.  For example, bash ./createKeystore.sh ...
#
#              The details needed by importToAgent.py include things like
#              the list of certificate alias' in the keystore as well as the
#              associated password. This script will use the certificate file
#              name (less extension) as the alias and a default password defined
#              in the importToAgentEnv.sh (AGENT_KS_PASSWORD). These lists are
#              captured in the importToAgent.ini and look something like the
#              following:
#
#              keyAliases: root-VeriSignClass3-G5,intermediate-Symantec-G4,main-us
#              keyPasswords: changeit,changeit,changeit
#
#===============================================================================

. ./importToAgentEnv.sh

##################################################
# Process arguments
##################################################
for i in "$@"
do
case $i in
    -cd=*|--certificateDir=*)
    CERTS_DIR="${i#*=}"
    shift
    ;;
    -cp=*|--certificatePattern=*)
    CERTS_PATTERN="${i#*=}"
    shift
    ;;
    -jh=*|--javaHome=*)
    JAVA_HOME="${i#*=}"
    shift
    ;;
    --help)
    HELP=YES
    ;;
    *)
    # unknown option
    ;;
esac
done


##################################################
# Expand the use of ~ in the javaHome
##################################################
JAVA_HOME_EXPANDED=$(eval echo $JAVA_HOME)


##################################################
# Setup Variables
##################################################
export KEYTOOL=$JAVA_HOME_EXPANDED/bin/keytool
export KEYSTORE=$CERTS_DIR/$JKS_FILENAME


##################################################
# Make sure the keytool is available
##################################################
if [[ ! -f $KEYTOOL ]]; then
    echo "\"$KEYTOOL\" does not exist."
    echo "Please verify JAVA_HOME in importToAgentEnv.sh"
    echo "or override using -jh command-line option, exiting ..."
    exit 1
fi


##################################################
# Check for usage
##################################################
if [[ ( -z $CERTS_DIR ) || ( -z $CERTS_PATTERN ) || ( "$HELP" = "YES" ) ]]; then
    echo "usage: ./$0 -cd=PathToCertificates -cp=CertificateFilePattern [-jh=JavaJDKHome]"
    echo "For example, ./$0 -cd=~/certificates -cp=*.crt"
    exit 1
fi


##################################################
# Check for valid certificates directory
##################################################
if [[ ! -d $CERTS_DIR ]]; then
    echo "Directory \"$CERTS_DIR\" does not exist, exiting ..."
    exit 1
fi


##################################################
# Check for certificate files
##################################################
CERTS=$(ls $CERTS_DIR/$CERTS_PATTERN | xargs -n1 basename)

if [[ -z $CERTS ]]; then
    echo "No certificates found for \"$CERTS_DIR/$CERTS_PATTERN\", exiting ..."
    exit 1
fi


##################################################
# Check for existing keystore and remove
##################################################
if [[ -f $KEYSTORE ]]; then
    echo "Keystore \"$KEYSTORE\" already exists, removing ..."
    rm $KEYSTORE
fi


##################################################
# Loop through certificates and add to keystore
# and create variables for wlst ini file
##################################################
echo "Certificates will be added to $KEYSTORE"

ALIASES=""
KEYPASSES=""

for c in $CERTS
do
    ALIAS=$(echo $c | cut -f 1 -d '.')
    KEYTOOL_CMD="$KEYTOOL -importcert -trustcacerts -keystore $KEYSTORE -storepass $AGENT_KS_PASSWORD -keypass $AGENT_KS_PASSWORD -storetype JKS -alias $ALIAS -file $CERTS_DIR/$c"
    echo "Adding certificate $c"
    $KEYTOOL_CMD
    echo ""
    
    if [[ ! -z $ALIASES ]]; then
        ALIASES=$ALIASES","
        KEYPASSES=$KEYPASSES","
    fi
    
    ALIASES=$ALIASES$ALIAS
    KEYPASSES=$KEYPASSES$AGENT_KS_PASSWORD
done

##################################################
# Create importToAgent.ini
##################################################
echo "[ImportKeyStore]"                     >  importToAgent.ini
echo "appStripe: system"                    >> importToAgent.ini
echo "keystoreName: trust"                  >> importToAgent.ini
echo "keyAliases: $ALIASES"                 >> importToAgent.ini
echo "keyPasswords: $KEYPASSES"             >> importToAgent.ini
echo "keystorePassword: $AGENT_KS_PASSWORD" >> importToAgent.ini
echo "keystorePermission: true"             >> importToAgent.ini
echo "keystoreType: JKS"                    >> importToAgent.ini
echo "keystoreFile: $KEYSTORE"              >> importToAgent.ini


##################################################
# List all the certificates in the keystore
##################################################
$KEYTOOL -list -keystore $KEYSTORE -storepass $AGENT_KS_PASSWORD

echo ""
echo ""
echo ""
echo "Keystore ready for connectivity agent import: $KEYSTORE"


advancedConfigScripts/importToAgentEnv.sh
#!/bin/bash
#===============================================================================
# Title:       importToAgentEnv.sh
#
# Disclaimer:  This script is provided "AS IS" and without any official support
#              from Oracle. Its use needs to be performed using details from the
#              comments section and/or readme file (if one is included). Any bugs
#              encountered, feedback, and/or enhancement requests are welcome.
#
#              No liability for the contents of this script can be accepted. Use
#              the concepts, examples, and information at your own risk. However,
#              great care has been taken to ensure that all technical information
#              is accurate and as useful as possible.
#
#              Copyright © 2018, Oracle and/or its affiliates. All rights reserved.
#              The Universal Permissive License (UPL), Version 1.0
#
#
# Description: This is the environment file for the ICS Connectivity Agent
#              createKeystore.sh and importToAgent.sh scripts.  The contents of
#              this file should be updated to reflect your environment.
#
# Author:      Greg Mally
#
# Date:        31-October-2016
#
# Version:     1.0    
#
# Usage:       . ./importToAgentEnv.sh
#
# Notes:       The file name for JKS_FILENAME will be created in the same
#              directory where the certificates are located.  If this file
#              already exists, it will be removed prior to importing the
#              certificates.
#
#              The AGENT_KS_PASSWORD is set to a standard default value of
#              "changeit".  This value shouldn't be modified unless you know
#              exactly what you are doing.
#
#===============================================================================

export JAVA_HOME=/oracle/java/jdk1.7.0

export AGENT_HOME=/oracle/agenthome

export JKS_FILENAME=agentcerts.jks

# The following (changeit) is the default for the connectivity agent keystore
# DO NOT CHANGE
export AGENT_KS_PASSWORD=changeit


credentialScripts/updateicscredentials.py
#=========================================================================================
# Title:       updateicscredentials.py - ICS Connectivity Agent credential update
#
# Disclaimer:  This script is provided "AS IS" and without any official support
#              from Oracle. Its use needs to be performed using details from the
#              comments section and/or readme file (if one is included). Any bugs
#              encountered, feedback, and/or enhancement requests are welcome.
#
#              No liability for the contents of this script can be accepted. Use
#              the concepts, examples, and information at your own risk. However,
#              great care has been taken to ensure that all technical information
#              is accurate and as useful as possible.
#
#              Copyright © 2018, Oracle and/or its affiliates. All rights reserved.
#              The Universal Permissive License (UPL), Version 1.0
#
#
# Description: This script is used to update the credentials for a running Connectivity
#              Agent server. It is necessary to run this script whenever the ICS user
#              that was used during agent installation has changed their password. If this
#              is not done, the agent will appear to be running fine but the ICS console
#              will show the agent as down. Additionally, if the ICS user who used their
#              credentials for the agent installation tries to login to ICS they will be
#              temporarily locked out. This is due to the agent making multiple calls to
#              ICS with invalid credentials resulting in ICS locking out the user.
#
#
# Usage:       $AGENT_HOME/oracle_common/common/bin/wlst.sh $PATH_TO_SCRIPT/updateicscredentials.py adminUser adminPass agentT3URI icsUser icsPass
#
# Example:     ./wlst.sh /tmp/updateicscredentials.py weblogic welcome1 t3://localhost:7001 ics.user@example.com N3wP@55w0rd
#
# NOTES:       It is important that the agent server be running when this script is
#              executed.
#
#              After executing the script, shut the agent server down and wait for
#              approximately 30 minutes to allow the locked user account to unlock
#              before attempting to start the agent.
#
#=========================================================================================

import os
import sys
import shutil


adminUserName=sys.argv[1]
adminPassword=sys.argv[2]
adminURL=sys.argv[3]
icsUserName=sys.argv[4]
icsPassword=sys.argv[5]


try:
    connect(adminUserName, adminPassword, adminURL)
    updateCred(map="oracle.agent.patcher", key="AGENT_PATCHER_APPID", user=icsUserName, password=icsPassword, desc="Updating Agent Credential Store")
    disconnect()
except:
   print "Error updating credentials for AGENT_PATCHER_APPID."
exit()


advancedConfigScripts/importToAgent.sh
#!/bin/bash
#===============================================================================
# Title:       importToAgent.sh
#
# Disclaimer:  This script is provided "AS IS" and without any official support
#              from Oracle. Its use needs to be performed using details from the
#              comments section and/or readme file (if one is included). Any bugs
#              encountered, feedback, and/or enhancement requests are welcome.
#
#              No liability for the contents of this script can be accepted. Use
#              the concepts, examples, and information at your own risk. However,
#              great care has been taken to ensure that all technical information
#              is accurate and as useful as possible.
#
#              Copyright © 2018, Oracle and/or its affiliates. All rights reserved.
#              The Universal Permissive License (UPL), Version 1.0
#
#
# Description: This script is used to validate and kick off the WebLogic WLST
#              importToAgent.py script.  It relies on the importToAgentEnv.sh
#              and will check for the following:
#                1. Is the Connectivity Agent running
#                2. Check for importToAgent.ini (generated by createKeystore.sh)
#                3. Check for WLST script in $AGENT_HOME/oracle_common/common/bin
#
#              If no command-line arguments are provided, the following defaults
#              will be used:
#                * Agent Server User:          weblogic
#                * Agent Server User Password: welcome1
#                * Agent Server Host Name:     localhost
#                * Agent Server Host Port:     7001
#
# Author:      Greg Mally
#
# Date:        18-May-2018
#
# Version:     1.1
#
# Usage:       ./importToAgent.sh -au=agentAdmin -ap=admin! -ah=agent.demo.org -aport=7101
#
# Notes:       The above usage assumes that this script has execute permission.
#              If it doesn't have execute permission, then add "bash" in front
#              of the command.  For example, bash ./importToAgent.sh ...
#
#              After this script validates the command-line arguments and the
#              various points in the description above, it will kick off WLST
#              with the importToAgent.py and provide the agent user name,
#              password, host name, and host port.
#
#              After running this script it is recommended to bounce the
#              Connectivity Agent server to make sure everything loads as
#              expected.
#
# History:     1.0 - Initial release on 31-October-2016
#
#              1.1 - Added function for inputting the agent password in a
#                    masked text fashion.
#
#                    Removed the -ap/--agentpassword option
#
#===============================================================================

. ./importToAgentEnv.sh

##################################################
# Default values
##################################################
AGENTUSER=weblogic
AGENTPASSWORD=welcome1
AGENTHOST=localhost
AGENTPORT=7001


##################################################
# Function: Read secret string
# Version 1.1 - Added this function
##################################################
function read_secret {
    # Disable echo
    stty -echo
    
    # Set up trap to ensure echo is enabled before exiting if the script
    # is terminated while echo is disabled.
    trap 'stty echo' EXIT
    
    # Read secret
    read SECRET
    
    # Enable echo and remove trap
    stty echo
    trap - EXIT
    
    # echo the secret to be captured in a bash variable
    echo $SECRET
}


##################################################
# Process arguments
# Version 1.1 - Remove agent password
##################################################
for i in "$@"
do
case $i in
    -au=*|--agentuser=*)
    AGENTUSER="${i#*=}"
    shift
    ;;
    -ah=*|--agenthost=*)
    AGENTHOST="${i#*=}"
    shift
    ;;
    -aport=*|--agentport=*)
    AGENTPORT="${i#*=}"
    shift
    ;;
    --help)
    HELP=YES
    ;;
    *)
    # unknown option
    ;;
esac
done


##################################################
# Check for usage
# Version 1.1 - Remove agent password
##################################################
if [[ ( -z $AGENTUSER ) || ( -z $AGENTHOST ) || ( -z $AGENTPORT ) || ( "$HELP" = "YES" ) ]]; then
    echo "Usage:       ./$0 [-au=AgentUsername] [-ah=AgentServerHost] [-aport=AgentServerPort]"
    echo "Defaults:    ./$0 -au=$AGENTUSER -ap=$AGENTPASSWORD -ah=$AGENTHOST -aport=$AGENTPORT"
    echo "For example, ./$0 -au=agentAdmin -ap=admin! -ah=agent.demo.org -aport=7101"
    exit 1
fi


##################################################
# Setup Variables
##################################################
export AGENT_HTTP_URL=http://$AGENTHOST:$AGENTPORT
export WLST=$AGENT_HOME/oracle_common/common/bin/wlst.sh


#################################################
# Check for Connectivity Agent Running
##################################################
wget -q $AGENT_HTTP_URL/console -O /dev/null

if [[ $? -ne 0 ]]; then
    echo "Agent is not running at: $AGENT_HTTP_URL"
    echo "Please start the Agent and try again."
    exit 1
fi


##################################################
# Check for importToAgent.ini that is generated
# by the createKeystore.sh script
##################################################
if [[ ! -f importToAgent.ini ]]; then
   echo "WLST .ini not found: importToAgent.ini"
   echo "Please generate using createKeystore.sh"
   exit 1
fi


##################################################
# Check for wlst
##################################################
if [[ ! -f $WLST ]]; then
   echo "WLST not found: $WLST"
   echo "Please verify Agent home: $AGENT_HOME"
   echo "The AGENT_HOME can be found in the importToAgentEnv.sh"
   exit 1
fi


##################################################
# Get Agent password
# Version 1.1 - Added agent password read
##################################################
# Display prompt, get password, and echo newline
printf "Agent Password: "
AGENTPASSWORD="$(read_secret)"
echo


##################################################
# Start the import process
##################################################
$WLST ./importToAgent.py -u $AGENTUSER -p $AGENTPASSWORD -h $AGENTHOST -t $AGENTPORT
